contracts/Elections.sol

```diff
@@ -273,24 +299,33 @@ contract Elections is IElections, ContractRegistryAccessor {
 		return bannedValidators[addr] != 0;
 	}
 
-	function notifyStakeChange(address stakeOwner, uint256 newUncappedStake, uint256 prevGovStakeOwner, address delegatee, uint256 prevGovStakeDelegatee) external onlyDelegationsContract {
-		_applyDelegatedStake(delegatee, newUncappedStake);
+	function notifyStakeChange(uint256 prevDelegateTotalStake, uint256 newDelegateTotalStake, address delegate, bool isSelfDelegatingDelegate) external onlyDelegationsContract {
 
-		_applyStakesToBanningBy(delegatee, prevGovStakeDelegatee);
+		uint256 prevGovStakeDelegate = getGovernanceEffectiveStake(isSelfDelegatingDelegate, prevDelegateTotalStake);
+		uint256 newGovStakeDelegate = getGovernanceEffectiveStake(isSelfDelegatingDelegate, newDelegateTotalStake);
+
+		totalGovernanceStake = totalGovernanceStake.sub(prevGovStakeDelegate).add(newGovStakeDelegate);
+
+		_applyDelegatedStake(delegate, newDelegateTotalStake);
+
+		_applyStakesToBanningBy(delegate, prevGovStakeDelegate);
 	}
 
-	function notifyStakeChangeBatch(address[] calldata stakeOwners, uint256[] calldata newUncappedStakes, uint256[] calldata prevGovStakeOwners, address[] calldata delegatees, uint256[] calldata prevGovStakeDelegatees) external onlyDelegationsContract {
-		require(stakeOwners.length == newUncappedStakes.length, "arrays must be of same length");
-		require(stakeOwners.length == prevGovStakeOwners.length, "arrays must be of same length");
-		require(stakeOwners.length == delegatees.length, "arrays must be of same length");
-		require(stakeOwners.length == prevGovStakeDelegatees.length, "arrays must be of same length");
+	function notifyStakeChangeBatch(uint256[] calldata prevDelegateTotalStakes, uint256[] calldata newDelegateTotalStakes, address[] calldata delegates, bool[] calldata isSelfDelegatingDelegates) external onlyDelegationsContract {
+		require(prevDelegateTotalStakes.length == newDelegateTotalStakes.length, "arrays must be of same length");
+		require(prevDelegateTotalStakes.length == delegates.length, "arrays must be of same length");
+		require(prevDelegateTotalStakes.length == isSelfDelegatingDelegates.length, "arrays must be of same length");
+
+		for (uint i = 0; i < prevDelegateTotalStakes.length; i++) {
+			uint256 prevGovStakeDelegate = getGovernanceEffectiveStake(isSelfDelegatingDelegates[i], prevDelegateTotalStakes[i]);
+			uint256 newGovStakeDelegate = getGovernanceEffectiveStake(isSelfDelegatingDelegates[i], newDelegateTotalStakes[i]);
 
-		for (uint i = 0; i < stakeOwners.length; i++) {
+			totalGovernanceStake = totalGovernanceStake.sub(prevGovStakeDelegate).add(newGovStakeDelegate);
```

Comment:
There are a lot of state writes here that seem redundant. They are probably needed for `_applyDelegatedStake` and `_applyStakesToBanningBy`, but maybe this could be avoided somehow.
`applyStakesToBanningBy` can probably run in a second iteration. Not sure about `_applyDelegatedStake`.

Reply:
we can edit totalGovernanceStake in memory, and pass the intermediary total into _applyDelegatedStake.

then write the state only once after this loop